SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityDeployChecker.java

```diff
@@ -319,6 +335,106 @@ private void finishDeploy(SingularityRequestWithState requestWithState, Optional
+                                                                  SingularityDeployResult deployResult,
+                                                                  Optional<SingularityRunNowRequest> maybeRunNowRequest) {
+    String requestId = request.getId();
+    String deployId = pendingDeploy.getDeployMarker().getDeployId();
+    Optional<String> user = pendingDeploy.getDeployMarker().getUser();
+    long timestamp = deployResult.getTimestamp();
+    PendingType pendingType;
+    Optional<Boolean> skipHealthChecks = request.getSkipHealthchecks();
+    Optional<String> message = pendingDeploy.getDeployMarker().getMessage();
+    Optional<List<String>> commandLineArgs;
+    Optional<String> runId;
+    Optional<Resources> resources;
+    List<SingularityTaskId> activeTasks = taskManager.getActiveTaskIdsForRequest(requestId);
+    List<SingularityPendingTaskId> pendingTasks = taskManager.getPendingTaskIdsForRequest(requestId);
+
+    SingularityPendingRequest pendingRequest;
+
+    if (request.isScheduled()
+        && maybeRunNowRequest.isPresent()
+        && activeTasks.isEmpty()) {
+      SingularityRunNowRequest runNowRequest = maybeRunNowRequest.get();
+      runId = runNowRequest.getRunId().or(Optional.of(UUID.randomUUID().toString()));
+      message = runNowRequest.getMessage()
+          .or(message);
+      commandLineArgs = runNowRequest.getCommandLineArgs();
+      skipHealthChecks = runNowRequest.getSkipHealthchecks().or(skipHealthChecks);
+      pendingType = PendingType.IMMEDIATE;
+      resources = runNowRequest.getResources();
+      pendingRequest = new SingularityPendingRequest(
+          requestId,
+          deployId,
+          timestamp,
+          user,
+          pendingType,
+          commandLineArgs,
+          runId,
+          skipHealthChecks,
+          message,
+          Optional.absent(),
+          resources);
+      return Optional.of(pendingRequest);
+    } else if (request.isScheduled()) {
+      // There is already a running task for this request. Don't attempt to run now, just redeploy
+      pendingType = deployResult.getDeployState() == DeployState.CANCELED
+          ? PendingType.DEPLOY_CANCELLED
+          : PendingType.NEW_DEPLOY;
+      pendingRequest = new SingularityPendingRequest(
+          requestId,
+          deployId,
+          timestamp,
+          user,
+          pendingType,
+          skipHealthChecks,
+          message);
+      return Optional.of(pendingRequest);
+    } else if (!request.isAlwaysRunning()
+        && maybeRunNowRequest.isPresent()
+        && request.getInstances().isPresent()
+        && (activeTasks.size() + pendingTasks.size() < request.getInstances().get())) {
+      SingularityRunNowRequest runNowRequest = maybeRunNowRequest.get();
+      pendingType = PendingType.ONEOFF;
+      runId = runNowRequest.getRunId().or(Optional.of(UUID.randomUUID().toString()));
+      message = runNowRequest.getMessage()
```

Comment:
might be going a bit line-break happy here/above/below in this method ;) can get a bit hard to read